Add cpo property Bingham average

MFraters opened the pull request:
Adds a particle property plugin to compute the Bingham average for the CPO particle properties.

This implementation is based on https://github.com/geodynamics/aspect/pull/4978, so this should not be merged until that one is. It is ready for review though.

(point 4 in https://github.com/geodynamics/aspect/issues/3885)

MFraters commented:
Now that #4978 is merged, I have rebased this pull request to the current main branch. It should now be fully ready.

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+      ASPECT_REGISTER_PARTICLE_PROPERTY(CpoBinghamAverage,
+                                        "cpo bingham average",
+                                        "This is a particle property plugin which computes the Bingham "
+                                        "average for the Crystal Preferred Orienation particle property "
```
```suggestion
                                        "average for the Crystal Preferred Orientation particle property "
```

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _aspect_particle_property_cpo_s_wave_anisotropy_h
```
please unify the name of the plugin and the name here.

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+           * Initialization function. This function is called once at the
+           * beginning of the program after parse_parameters is run.
+           */
+          virtual
+          void
+          initialize ();
```
```suggestion
          void
          initialize () override;
```
Almost the same, but better practice that we already do in all other plugins (yours were probably forked before we made the change). See here for a discussion why if you are interested: https://github.com/geodynamics/aspect/issues/3258

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+           * of this function should be to extend this vector by a number of
+           * properties.
+           */
+          virtual
```
as above

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+           * @param [in,out] particle_properties The properties of the particle
+           * that is updated within the call of this function.
+           */
+          virtual
```
as above, and for all other functions in this header that are marked as `virtual`

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+          get_property_information() const;
+
+          /**
+           * Computes the Bingham average
```
here would be a good place to briefly explain what a Bingham average is. You can also link to other places, but a short explanation should be here.

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+          /**
+          * Get volume weighted euler angles, using random draws to convert
+          * to a discrete number of orientations, weighted by volume.
+           */
```
```suggestion
          */
```

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int cpo_data_position;
```
missing documentation

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+        private:
+          unsigned int cpo_data_position;
+
+          double rad_to_degree = 180.0/M_PI;
```
Let's not start another way to define pi, we already have one:
```suggestion
          constexpr double rad_to_degree = 180.0/numbers::PI;
```

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+          unsigned int cpo_data_position;
+
+          double rad_to_degree = 180.0/M_PI;
+          double degree_to_rad = M_PI/180.0;
```
```suggestion
          constexpr double degree_to_rad = numbers::PI/180.0;
```

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,202 @@
+          unsigned int n_grains;
+          unsigned int n_minerals;
+
+          // when doing the random draw volume weighting, this sets how many samples are taken.
```
```suggestion
          /**
           * when doing the random draw volume weighting, this sets how many samples are taken.
           */
```

gassmoeller reviewed (commented):
Only comments on the header so far, have to board a plane. Rest to come.

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+        unsigned int n_grain = fv_to_sort.size();
+
+
+        /**
+         * ...
+         */
```
This needs documentation. Also your algorithm here is N^2 in the size of the vectors (linear because of your loop and linear because of max_element) and quite expensive per step, because you sort both vectors. What is the typical size of these vectors? If they are bigger than 100 I would consider a better algorithm, e.g. the one suggested in this answer on stackoverflow: https://stackoverflow.com/a/17074810 (you can replace the `compare` template with <). Also you do not really need to sort the matrices, you just need to keep a permutation from the sorted to the unsorted array (which this other algorithm gives you as byproduct). That way you can keep the vector of matrices untouched.

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+        // the euler angle index is +1. For example, if the idxGrain(g) < cumWeight(1),
+        // the index should be 1 not zero)
+        std::vector<Tensor<2,3>> matrices_out(n_output_grains);
+        for (unsigned int grain_i = 0; grain_i < n_output_grains; ++grain_i)
```
Check https://github.com/dealii/dealii/blob/38356ce009ba1b041bc97655d134bccd2480d8d1/source/particles/generators.cc#L389, this should make this algorithm much simpler (and NlogN instead of N^2).

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+      CpoBinghamAverage<dim>::get_property_information() const
+      {
+        std::vector<std::pair<std::string,unsigned int>> property_information;
+        for (size_t mineral_i = 0; mineral_i < n_minerals; mineral_i++)
```
here `size_t` is the wrong type. `unsigned int` will do.

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+
+      template<int dim>
+      std::vector<Tensor<2,3>>
+      CpoBinghamAverage<dim>::random_draw_volume_weighting(std::vector<double> fv,
```
what is fv? volume_fraction? if so, please rename

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+            unsigned int counter = 0;
+            for (unsigned int grain_j = 0; grain_j < n_grains; ++grain_j)
+              {
+                // find the first cummulative weight which is larger than the random number
```
```suggestion
                // find the first cumulative weight, which is larger than the random number
```

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,369 @@
+      {
+        // Get volume weighted euler angles, using random draws to convert odf
+        // to a discrete number of orientations, weighted by volume
+        // 1a. Get index that would sort volume fractions AND
```
I think here some documentation is missing, the next line seems out of place.

gassmoeller reviewed (commented):
Some comments on the implementation. Most are small and quick to fix. I have a few suggestions on optimizations for the algorithms, but you need to consider how important performance in this part is (compared to the other parts of CPO) to check if it is worth it to improve the algorithms.

MFraters commented:
@gassmoeller Thanks for the review. I have addressed your comments and replaced the random draw function with a new utilities one (see #5094) based on your suggestion.

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,239 @@
+
+          /**
+           * Create a permutation vector which can be used by the apply_permutation function
+           * which shorts the vector in .. order.
```
which order?

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,239 @@
+            const std::vector<T> &vector) const;
+
+          /**
+           * Applies a permutation vector to return a sorted verions of the vector.
```
```suggestion
           * Applies a permutation vector to return a sorted version of the vector.
```

gassmoeller commented on include/aspect/particle/property/cpo_bingham_average.h:
```diff
@@ -0,0 +1,239 @@
+          Tensor<3,3> permutation_operator_3d;
+
+          // utility values
+          double rad_to_degree = 180.0/M_PI;
+          double degree_to_rad = M_PI/180.0;
```
we now have constants for this, see #5100.

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,329 @@
+        permutation_operator_3d[1][0][2]  = -1;
+        permutation_operator_3d[2][1][0]  = -1;
+      }
+
```
three empty lines between function definitions

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,329 @@
+          }
+        };
+      }
+
```
empty lines between functions, check here and all other functions.

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,329 @@
+                                 "The seed used to generate random numbers. This will make sure that "
+                                 "results are reproducable as long as the problem is run with the "
+                                 "same amount of MPI processes. It is implemented as final seed = "
+                                 "user seed + MPI Rank. ");
```
```suggestion
                                 "Random number seed + MPI Rank. ");
```

gassmoeller commented on source/particle/property/cpo_bingham_average.cc:
```diff
@@ -0,0 +1,329 @@
+        {
+          prm.enter_subsection("Particles");
+          {
+            prm.enter_subsection("CpoBinghamAverage");
```
To be consistent with our usual naming scheme of subsection this would need to  be:
```suggestion
            prm.enter_subsection("CPO Bingham Average");
```

gassmoeller reviewed (commented):
Looks good, just some minor comments left. I guess we need #4979 before we can test this?

gassmoeller reviewed (approved):
Looks good, ready to go. I saw a few minor things that I will address in a follow-up. I will ping you in that PR.
